### ISSUE TITLE
Dangling false positive if the the owner is also moved in the initializer.

### ISSUE THREAD

**Issue by hokein** [llvm/llvm-project#114201]
This issue is identified in #112751.

```cpp
// case1
namespace std {
template<typename T>
struct unique_ptr {
  T &operator*();
  T *get() const [[clang::lifetimebound]];
};
} // namespace std

struct X {
  X(std::unique_ptr<int> up) :
    pointer(up.get()), owner(std::move(up)) {}

  int *pointer;
  std::unique_ptr<int> owner;
};
```

When we add the `clang::lifetimebound` annotation to `unique_ptr::get()`, clang emits a dangling-field warning for the `pointer(up.get())` member initializer. This warning is a false positive in this context, as the `owner` member is moved as part of the initialization, retaining ownership.

Another example occurs in designated-initializer cases:

```cpp
// case2
struct X {
   int *pointer;
   std::unique_ptr<int> owner;
};

X func(std::unique_ptr<int> up) {
   return {
      .pointer = up.get(),
      .owner = std::move(up)
   };
}
```

Fixing these false positives is hard because it would require tracking dependencies between expressions, which is beyond the capabilities of the current statement-local analysis.

**Comment by hokein** [comment]
Interestingly, clang will emit the diagnostic on case 2 even the `get()` is not annotated with `lifetimebound`, https://godbolt.org/z/vYeG9Eb3Y.

The GSL analysis implicitly tracks the `*this` object for some hard-coded GSL owner methods (`vector.get()`), the case 1 have been filtered out [here](https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/CheckExprLifetime.cpp#L1264-L1267), but it is just for the GSL code path.
